### DISCUSSION TITLE
Reimplementing the runtime in C#

### DISCUSSION THREAD

**Discussion by colejohnson66** [dotnet/runtime#71442] [General] [answered]
Crazy thought idea, but would it even be possible for the runtime be written in C#? I know the C# compiler was rewritten into C# (Roslyn), but the runtime is still not self-hosted.

The first showstopper I can think of would be memory allocation and the garbage collector. Basically, a self-hosted runtime would need a way to allocate memory and clean it up. So the self-hosted runtime would need its own runtime below it. Runtimes all the way down.

But now that AOT compilation exists, could that be worked around at all?

**Comment by colejohnson66** [comment]
There's also the issue of P/Invoke. You can't P/Invoke without a runtime, but we *are* the runtime. It would possibly require modifying the AOT compilation step to detect things like P/Invoke and turn them into native DLL stub calls. Any allocations would require P/Invoking to `malloc` and `free` with unsafe pointers.

**Comment by stephentoub** [comment, marked as answer]
https://github.com/dotnet/runtime/issues/10158

**Reply by colejohnson66** [reply]
Interesting. The first thing I notice is that the discussion is from 2018, which is before AOT compilation was released.
